### ISSUE TITLE
[TFLite] Accumulator and bias types coherence for int16x8 FC operator

### ISSUE THREAD

**Issue by Tessil** [tensorflow/tensorflow#53763]
Hello,

We noticed that int32 accumulator and int32 bias support was recently added for the int16x8 FULLY_CONNECTED operator (along with CONV_2D and TRANSPOSE_CONV_2D operators) through the `_experimental_full_integer_quantization_bias_type` option of the TFLiteConverter :
* [Add option to store bias as 32 bit in 16x8 Quant.](https://github.com/tensorflow/tensorflow/commit/ea33c1e7a25d8025e8ee405ad8ab7be261798d76)
* [Update TFLite kernel to use Ruy 16x8 Gemm instead of reference kernel.](https://github.com/tensorflow/tensorflow/commit/5de18207d9568ee25bdd57db8eaf6268fe0b913d) 

It seems though that these changes are leading to some incoherences with potential unexpected overflow and also created a bug. Before an int16x8 FC would always use an int64 accumulator to avoid any overflow and an int64 bias. The current status now seems to be:
* converter._experimental_full_integer_quantization_bias_type = tf.int64 or None:
    * OpResolverType.BUILTIN_REF
        * use_bias = True
            * int64 accumulator and int64 bias
        * use_bias = False
            * int64 accumulator
    * OpResolverType.BUILTIN
        * use_bias = True
            * int64 accumulator and int64 bias
        * use_bias = False
            * int32 accumulator (would have expected an int64 accumulator)
 * converter._experimental_full_integer_quantization_bias_type = tf.int32:
    * OpResolverType.BUILTIN_REF
        * use_bias = True
            * int64 accumulator and int64 bias (read an int32 bias tensor as an int64 tensor => bug, need to be an int32 bias and would expect an int32 accumulator)
        * use_bias = False
            * int64 accumulator (would have expected an int32 accumulator)
    * OpResolverType.BUILTIN
        * use_bias = True
            * int32 accumulator and int32 bias
        * use_bias = False
            * int32 accumulator

Ideally we would have expected that an int64 accumulator (no matter if the layer has no bias) and int64 bias would be used when `converter._experimental_full_integer_quantization_bias_type = tf.int64 or None` (both with the reference and optimized kernels) and an int32 accumulator (no matter if the layer has no bias) and int32 bias when `converter._experimental_full_integer_quantization_bias_type = tf.int32` (both with the reference and optimized kernels).

These incoherences also affect the CONV_2D and TRANPOSE_CONV_2D operators and we were wondering if it was intentional (and if so the relation between the bias type and the accumulator type) or just an oversight?

Code to illustrate the bug with `converter._experimental_full_integer_quantization_bias_type = tf.int32`, `OpResolverType.BUILTIN_REF` and `use_bias=True`. It's due to the https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L835 condition that will be true if `kernel_type == kReference` no matter the type of the bias while `FullyConnectedInt16` will always assume an int64 bias and thus read the int32 bias tensor as an int64 tensor. It fails to provide the expected results with tf_nightly-2.9.0.dev20220114.

```Python
import tensorflow as tf

input = tf.keras.Input(shape=[1], batch_size=1)
output = tf.keras.layers.Dense(
    10,
    kernel_initializer=tf.keras.initializers.Constant(value=0),
    bias_initializer=tf.keras.initializers.Constant(value=1),
)(input)
model = tf.keras.Model(inputs=[input], outputs=output)
model.summary()



def representative_dataset():
    yield [tf.constant([1.0])]

converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8
]
converter._experimental_full_integer_quantization_bias_type = tf.int32 # Will work if commented

tflite_model = converter.convert()
with open("test.tflite", "wb") as fp:
    fp.write(tflite_model)



interpreter = tf.lite.Interpreter(
    model_path="test.tflite",
    experimental_op_resolver_type=tf.lite.experimental.OpResolverType.BUILTIN_REF, # Will work if commented
)
interpreter.allocate_tensors()

input = [tf.constant([1.0])]
interpreter.set_tensor(interpreter.get_input_details()[0]['index'], input)
interpreter.invoke()
output = interpreter.get_tensor(interpreter.get_output_details()[0]['index'])

# [1.]
print(input)
# Expected: [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
print(output)
```

FYI @dayeongl

**Comment by mohantym** [comment]
Hi @Tessil! Thanks for the reporting the issue in 2.9 . Please switch to stable version 2.7 for a while. Thank you!

**Comment by dayeongl** [comment]
thanks for reporting! I will work on it

**Comment by Tessil** [comment]
@mohantym Would it be possible to re-open the issue? Thanks!

**Comment by mohantym** [comment]
Ok @Tessil ! Re-opening as it is still replicating in [2.9](https://colab.sandbox.google.com/drive/1bl87zJXGeWRShJVqTMx-zk7AVy7m8R66?resourcekey=0-qP8tlN_1-i1fbmuYZaMfyg#scrollTo=Gm5V3_HCiAQD) . Thank you!

**Comment by rewu93** [comment]
One fundamental question: do we really want accumulators to be 32bits?  16 bits activation X 8 bits weights FC means that the maximum dimension we can have before integer overflow is 2^8 = 256.

**Comment by thaink** [comment]
> One fundamental question: do we really want accumulators to be 32bits? 16 bits activation X 8 bits weights FC means that the maximum dimension we can have before integer overflow is 2^8 = 256.

I think that was considered when Dayeong added this feature at the first place. That is why we set the 64bit as the default to avoid overflowing. Users will need to check the accuray by themselves and set the `_experimental_full_integer_quantization_bias_type` if it works for them. From the description of that change:
"For 16x8 Quant, the default is to store the bias as 64 bit. However, 64 bit is 1.5~2 times slower than 32 bits so there is a feature request to enable 32 bit bias in 16x8 Quant. Users can test their model's accuracy and performance with both 32 bit and 64 bit and finalize which options to go by themselves. The default is {32bit for int8Quant, 64bit for int16Quant}."

**Comment by rewu93** [comment]
Understood. However, it it possible to have 64 bits accumulator but 32 bits bias? My concern is that integer overflow is really hard to debug.

**Comment by alankelly** [comment]
@abattery Could you please take a look at this? Thanks

**Comment by pkgoogle** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.

**Comment by Tessil** [comment]
The issue is still present, there's a [PR](https://github.com/tensorflow/tensorflow/pull/58400) to fix it. We should wait for it to be merged before closing the issue.

**Comment by pkgoogle** [comment]
Hi, @Tessil, it seems the corresponding PR has merged, can you confirm the issue is fixed?

**Comment by Tessil** [comment]
@pkgoogle Thanks, the PR was effectively merge but it seems there's still a bug remaining, see https://github.com/tensorflow/tensorflow/pull/58400#issuecomment-1723176172 comment.

**Comment by pkgoogle** [comment]
Hi @Tessil ,

Thanks for raising this issue. Are you aware of the migration to [LiteRT](https://developers.googleblog.com/en/tensorflow-lite-is-now-litert/)? This transition is aimed at enhancing our project's capabilities and providing improved support and focus for our users. As we believe this issue is still relevant to LiteRT we are moving your issue there. Please follow progress [here](https://github.com/google-ai-edge/LiteRT/issues/129).

Let us know if you have any questions. Thanks.
